packages/contracts/contracts/deploy/IPayableDAppSystem.sol

```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: Apache-2.0
+/// @title ERC20 Payment System factory
+pragma solidity ^0.8.13;
+
+import {IConsensus} from "@cartesi/rollups/contracts/consensus/IConsensus.sol";
+import {ICartesiDAppFactory} from "@cartesi/rollups/contracts/dapp/ICartesiDAppFactory.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import {IPayableDAppFactory} from "./IPayableDAppFactory.sol";
+
+/// @notice Interface for creating new ERC20 based DApp factories
+interface IPayableDAppSystem {
+    /// @notice A new factory was created
+    /// @param factory The address of the factory
+    /// @param token The token used for billing
+    /// @param consensus The consensus associated with the factory
+    /// @param price The price per minute of dapp node execution
+    event PayableDAppFactoryCreated(
+        IPayableDAppFactory factory,
+        IERC20 token,
+        IConsensus consensus,
+        uint256 price
+    );
+
+    /// @notice Create a new ERC20 based DApp factory using the specified token
+    function newPayableDAppFactory(
+        IERC20 _token,
+        IConsensus consensus,
+        uint256 _price
+    ) external returns (IPayableDAppFactory);
```

Comment:
Then, you'd define a new interface for it.
I like to type factories with the actual contracts instead of their interfaces, because then you generally have more control over the contracts that you've just deployed.
For example, imagine a factory that deploys `Ownable` contracts. If you type them by the interface, they will not have the functions from `Owanble` such as `transferOwnership`, which might be useful for factories that use other factories.

Reply:
`Ownable` is an abstract contract, not an interface.
If there was an interface, i.e. `IOwnable` I'd use it in my `IOwnableFactory`, and `transferOwnership` would be part of the interface rather than the contract.

Do you have another example?

I still think the correct is to return the interface.

Reply:
> you generally have more control over the contracts that you've just deployed

It shouldn't matter, as long as it implements the interface

Reply:
> If there was an interface, i.e. IOwnable I'd use it in my IOwnableFactory, and transferOwnership would be part of the interface rather than the contract.

The thing is, that there isn't such an interface.
OpenZeppelin only defines `Ownable`, but no `IOwnable`.
That is why I chose to make `CartesiDAppFactory` return `CartesiDApp`.

Reply:
> It shouldn't matter, as long as it implements the interface

It matters, if the contract contains functions that are not exposed by the interface.

Reply:
> It matters, if the contract contains functions that are not exposed by the interface.

The interface should expose all functions that make sense for the interface.

If you need a function of the contract you type cast it manually and use it.

Reply:
> The case of OpenZeppelin is not related to the case of CartesiDAppFactory

It is, because `CartesiDApp` inherits from `Ownable`.

Reply:
> The interface should expose all functions that make sense for the interface.

That is not always the case. Some contracts inherit from several interfaces.

Reply:
> Some contracts inherit from several interfaces.

True, so you cast to the interface you want to interact as.

> It is, because CartesiDApp inherits from Ownable.

So if you want a `ICartesiDApp` to always behave as an `Ownable` you should create a `transferOwnership` method in the `ICartesiDApp`.


What defines expected behavior is the interface, not the implementation of the interface.
That is the whole point of interfaces. Clients should be dealing with interfaces, not concrete classes.

That is how interfaces in general programming languages work, and I don't see the reason to be different in solidity. Maybe I'm missing something.

Reply:
> True, so you cast to the interface you want to interact as.

Yes, Solidity allows you to cast any address value to any contract type. But I find it weird that you're upcasting `Apple` to `Fruit`, exposing `Fruit` and then downcasting to `Apple` again. Why not expose `Apple` instead?

> So if you want a ICartesiDApp to always behave as an Ownable you should create a transferOwnership method in the ICartesiDApp.

The thing is that you can't, with the current state of Solidity. The compiler raises an error saying that the function has already been defined (in this case, in `Ownable`).

> Clients should be dealing with interfaces, not concrete classes.

I generally agree with this statement. But factories are a special case, because their whole job is to deploy actual, concrete contracts.

Reply:
> I find it weird that you're upcasting `Apple` to `Fruit`

I'm not upcasting, I'm just returning the general representation

> and then downcasting to `Apple` again.

I won't downcast to `Apple`.

> The thing is that you can't, with the current state of Solidity. The compiler raises an error saying that the function has already been defined (in this case, in `Ownable`).

This is unfortunate.